// nani/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromArray = exports.collapseInfo = exports.filter = exports.find = exports.iterateFull = exports.iterate = exports.getFullStack = exports.getFullName = exports.is = exports.MultiError = exports.NaniError = void 0;
var nani_error_1 = require("./nani-error");
Object.defineProperty(exports, "NaniError", { enumerable: true, get: function () { return nani_error_1.NaniError; } });
var multi_error_1 = require("./multi-error");
Object.defineProperty(exports, "MultiError", { enumerable: true, get: function () { return multi_error_1.MultiError; } });
var is_1 = require("./is");
Object.defineProperty(exports, "is", { enumerable: true, get: function () { return is_1.is; } });
var get_full_name_1 = require("./get-full-name");
Object.defineProperty(exports, "getFullName", { enumerable: true, get: function () { return get_full_name_1.getFullName; } });
var get_full_stack_1 = require("./get-full-stack");
Object.defineProperty(exports, "getFullStack", { enumerable: true, get: function () { return get_full_stack_1.getFullStack; } });
var iterate_external_1 = require("./iterate-external");
Object.defineProperty(exports, "iterate", { enumerable: true, get: function () { return iterate_external_1.iterate; } });
Object.defineProperty(exports, "iterateFull", { enumerable: true, get: function () { return iterate_external_1.iterateFull; } });
var find_external_1 = require("./find-external");
Object.defineProperty(exports, "find", { enumerable: true, get: function () { return find_external_1.find; } });
Object.defineProperty(exports, "filter", { enumerable: true, get: function () { return find_external_1.filter; } });
var collapse_info_1 = require("./collapse-info");
Object.defineProperty(exports, "collapseInfo", { enumerable: true, get: function () { return collapse_info_1.collapseInfo; } });
var from_array_1 = require("./from-array");
Object.defineProperty(exports, "fromArray", { enumerable: true, get: function () { return from_array_1.fromArray; } });
