// objection-paginator/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownSortError = exports.InvalidCursorError = exports.ConfigurationError = exports.SortDirection = exports.ColumnType = exports.Paginator = void 0;
var paginator_js_1 = require("./paginator.js");
Object.defineProperty(exports, "Paginator", { enumerable: true, get: function () { return paginator_js_1.Paginator; } });
var sort_descriptor_js_1 = require("./sort-descriptor.js");
Object.defineProperty(exports, "ColumnType", { enumerable: true, get: function () { return sort_descriptor_js_1.ColumnType; } });
Object.defineProperty(exports, "SortDirection", { enumerable: true, get: function () { return sort_descriptor_js_1.SortDirection; } });
var configuration_error_js_1 = require("./configuration-error.js");
Object.defineProperty(exports, "ConfigurationError", { enumerable: true, get: function () { return configuration_error_js_1.ConfigurationError; } });
var invalid_cursor_error_js_1 = require("./invalid-cursor-error.js");
Object.defineProperty(exports, "InvalidCursorError", { enumerable: true, get: function () { return invalid_cursor_error_js_1.InvalidCursorError; } });
var unknown_sort_error_js_1 = require("./unknown-sort-error.js");
Object.defineProperty(exports, "UnknownSortError", { enumerable: true, get: function () { return unknown_sort_error_js_1.UnknownSortError; } });
